runsc/cmd/update.go

```diff
@@ -0,0 +1,296 @@
+
+	f.StringVar(&u.memory, "memory", "", "Memory limit (in bytes)")
+	f.StringVar(&u.memoryReservation, "memory-reservation", "", "Memory reservation or soft_limit (in bytes)")
+	f.StringVar(&u.memorySwap, "memory-swap", "", "Total memory usage (memory + swap); set '-1' to enable unlimited swap")
+
+	f.Int64Var(&u.pidsLimit, "pids-limit", 0, "Maximum number of pids allowed in the container")
+
+	f.StringVar(&u.l3CacheSchema, "l3-cache-schema", "", "The string of Intel RDT/CAT L3 cache schema")
+	f.StringVar(&u.memBwSchema, "mem-bw-schema", "", "The string of Intel RDT/MBA memory bandwidth schema")
+}
+
+// Execute implements subcommands.Command.Execute.
+func (u *Update) Execute(_ context.Context, f *flag.FlagSet, args ...any) subcommands.ExitStatus {
+	if f.NArg() != 1 {
+		f.Usage()
+		return subcommands.ExitUsageError
+	}
+
+	id := f.Arg(0)
+	conf := args[0].(*config.Config)
+
+	c, err := container.Load(conf.RootDir, container.FullID{ContainerID: id}, container.LoadOpts{})
+	if err != nil {
+		util.Fatalf("loading container %v", err)
+	}
+
+	r := specs.LinuxResources{
+		Memory: &specs.LinuxMemory{
+			CheckBeforeUpdate: boolPtr(false),
+		},
+		CPU:     &specs.LinuxCPU{},
+		BlockIO: &specs.LinuxBlockIO{},
+		Pids:    nil,
+	}
+	if in := u.resources; in != "" {
+		var (
+			f   *os.File
+			err error
+		)
+		switch in {
+		case "-":
+			f = os.Stdin
+		default:
+			if f, err = os.Open(in); err != nil {
+				return util.Errorf("opening %q: %v", in, err)
+			}
+			defer f.Close()
+		}
+		if err := json.NewDecoder(f).Decode(&r); err != nil {
+			return util.Errorf("decoding %q: %v", in, err)
+		}
+	} else {
+		if u.l3CacheSchema != "" || u.memBwSchema != "" {
+			return util.Errorf("Intel RDT support is not yet implemented")
+		}
+
+		if u.blkioWeight != 0 {
+			r.BlockIO.Weight = u16Ptr(uint16(u.blkioWeight))
+		}
+
+		r.CPU.Cpus = u.cpusetCpus
+		r.CPU.Mems = u.cpusetMems
+
+		for _, pair := range []struct {
+			strval string
+			dest   **uint64
+		}{
+			{u.cpuBurst, &r.CPU.Burst},
+			{u.cpuPeriod, &r.CPU.Period},
+			{u.cpuRtPeriod, &r.CPU.RealtimePeriod},
+			{u.cpuShares, &r.CPU.Shares},
+		} {
+			if pair.strval == "" {
+				continue
+			}
+			v, err := strconv.ParseUint(pair.strval, 10, 64)
+			if err != nil {
+				return util.Errorf("invalid value for %s: %v", pair.strval, err)
+			}
+			*pair.dest = &v
+		}
+
+		for _, pair := range []struct {
+			strval string
+			dest   **int64
+		}{
+			{u.cpuIdle, &r.CPU.Idle},
+			{u.cpuQuota, &r.CPU.Quota},
+			{u.cpuRtRuntime, &r.CPU.RealtimeRuntime},
+		} {
+			if pair.strval == "" {
+				continue
+			}
+			v, err := strconv.ParseInt(pair.strval, 10, 64)
+			if err != nil {
+				return util.Errorf("invalid value for %s: %v", pair.strval, err)
+			}
+			*pair.dest = &v
+		}
+
+		for _, pair := range []struct {
+			strval string
+			dest   **int64
+		}{
+			{u.memory, &r.Memory.Limit},
+			{u.memoryReservation, &r.Memory.Reservation},
+			{u.memorySwap, &r.Memory.Swap},
```

Comment:
This loop is identical to the loop above. Just add these entries above and get rid of this loop.

Reply:
good catch, fixed!